### PR TITLE
feat(autoscaling): add support for `InstanceRequirements` property

### LINKED ISSUE #28393: AutoScalingGroup: LaunchTemplateOverrides is missing the InstanceRequirements Attribute in AWS CDK L2 construct
### Describe the feature

The LaunchTemplateOverridesProperty is a crucial feature for our use case as it allows us to specify CPU, and memory configurations on a per-instance basis within the Auto Scaling Group and define spot and on-deamnd instances capcaity percentage. While this functionality is available in the L1 construct, it seems to be missing in the L2 construct.

### Use Case

We expect the L2 construct (AutoScalingGroup) to provide support for the LaunchTemplateOverridesProperty, enabling us to define specific configurations (such as min and max CPU and memory) and spot and on-demand capacity percentage within the Auto Scaling Group.

### Proposed Solution

We propose extending the L2 AutoScalingGroup construct to include support for the LaunchTemplateOverridesProperty or providing an alternative mechanism to achieve per-instance configuration within the Auto Scaling Group.

### Other Information

_No response_

### Acknowledgements

- [x] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.100.0

### Environment details (OS name and version, etc.)

ubuntu

### PR DESCRIPTION
**Pull request by wafuwafu13**
Closes #28393

> Basically [LaunchTemplateOverrides](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_autoscaling.LaunchTemplateOverrides.html) for L2 construct is missing the [InstanceRequirements](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-launchtemplateoverrides.html#cfn-autoscaling-autoscalinggroup-launchtemplateoverrides-instancerequirements) attribute.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts [hunk 1]
```diff
@@ -550,13 +550,27 @@ export interface InstancesDistribution {
...
+   * Then, it uses your On-Demand and Spot allocation strategies to launch instances from these instance types.
+   * You can specify up to four separate sets of instance requirements per Auto Scaling group.
+   * This is useful for provisioning instances from different Amazon Machine Images (AMIs) in the same Auto Scaling group.
+   * To do this, create the AMIs and create a new launch template for each AMI.
+   * Then, create a compatible set of instance requirements for each launch template.
+   *
+   * If you specify InstanceRequirements, you can't specify InstanceType.
+   *
+   * @default - Do not override instance type
+   */
```
Comment (wafuwafu13) [inline packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:564]: ref: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-launchtemplateoverrides.html#cfn-autoscaling-autoscalinggroup-launchtemplateoverrides-instancerequirements

File: packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts [hunk 2]
```diff
@@ -1810,11 +1824,15 @@ export class AutoScalingGroup extends AutoScalingGroupBase implements
                 if (override.weightedCapacity && Math.floor(override.weightedCapacity) !== override.weightedCapacity) {
                   throw new Error('Weight must be an integer');
                 }
+                if (override.instanceType && override.instanceRequirements) {
+                  throw new Error('\'InstanceRequirements\' can\'t be specified with \'InstanceType\'');
+                }
```
Comment (wafuwafu13) [inline packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:1829]: ref: 
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-launchtemplateoverrides.html#cfn-autoscaling-autoscalinggroup-launchtemplateoverrides-instancerequirements
>If you specify InstanceRequirements, you can't specify InstanceType.

**Comment by wafuwafu13** [comment]
Exemption Request

> ❌ Features must contain a change to a README file.

I add doc comment for `instanceRequirements` property

> ❌ Features must contain a change to an integration test file and the resulting snapshot.

I add test at `auto-scaling-group.test.ts`

**Comment by wafuwafu13** [comment]
Exemption Request

I add doc comment for instanceRequirements property
I add test at auto-scaling-group.test.ts

**Comment by wafuwafu13** [comment]
Exemption Request: I add doc comment for instanceRequirements property / I add test at auto-scaling-group.test.ts

**Comment by wafuwafu13** [comment]
Exemption Request: I add doc comment for instanceRequirements property / I add test at auto-scaling-group.test.ts

**Comment by sumupitchayan** [comment]
@wafuwafu13 please add an integration test or update an existing one

File: packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts [hunk 3]
```diff
@@ -550,13 +550,27 @@ export interface InstancesDistribution {
...
+   * You can specify up to four separate sets of instance requirements per Auto Scaling group.
+   * This is useful for provisioning instances from different Amazon Machine Images (AMIs) in the same Auto Scaling group.
+   * To do this, create the AMIs and create a new launch template for each AMI.
+   * Then, create a compatible set of instance requirements for each launch template.
+   *
+   * If you specify InstanceRequirements, you can't specify InstanceType.
+   *
+   * @default - Do not override instance type
+   */
+  readonly instanceRequirements?: CfnAutoScalingGroup.InstanceRequirementsProperty
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:565]: We'll need to implement the interface in the L2 construct and not make use of the existing one from L1.

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Thanks for starting this.
This is still a work in progress as the property needs to be implemented in the L2 construct directly.
Also, the related documentation, unit, and integration tests are required as this is a new feature.

File: packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts (hunk 3 above)
Comment (wafuwafu13) [reply packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:565]: Do you mean I should define `instanceRequirements` in another place?
If so, I would be grateful if you could tell me where to define it.

File: packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts [hunk 4]
```diff
@@ -1810,11 +1824,15 @@ export class AutoScalingGroup extends AutoScalingGroupBase implements
                 if (override.weightedCapacity && Math.floor(override.weightedCapacity) !== override.weightedCapacity) {
                   throw new Error('Weight must be an integer');
                 }
+                if (override.instanceType && override.instanceRequirements) {
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:1827]:
```suggestion
                if (!override.instanceType && !override.instanceRequirements) {
                  throw new Error('You must specify either \'instanceRequirements\' or \'instanceType\'.');
                }
                if (override.instanceType && override.instanceRequirements) {
```
I think we should validate that one of the two properties is specified (will require unit test coverage).

File: packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts [hunk 5]
```diff
@@ -550,13 +550,27 @@ export interface InstancesDistribution {
...
 export interface LaunchTemplateOverrides {
+  /**
+   * The instance requirements. Amazon EC2 Auto Scaling uses your specified requirements to identify instance types.
+   * Then, it uses your On-Demand and Spot allocation strategies to launch instances from these instance types.
+   * You can specify up to four separate sets of instance requirements per Auto Scaling group.
+   * This is useful for provisioning instances from different Amazon Machine Images (AMIs) in the same Auto Scaling group.
+   * To do this, create the AMIs and create a new launch template for each AMI.
+   * Then, create a compatible set of instance requirements for each launch template.
+   *
+   * If you specify InstanceRequirements, you can't specify InstanceType.
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:561]:
```suggestion
   * You must specify one of instanceRequirements or instanceType.
```

File: packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts [hunk 6]
```diff
@@ -550,13 +550,27 @@ export interface InstancesDistribution {
...
+   *
+   * @default - Do not override instance type
+   */
+  readonly instanceRequirements?: CfnAutoScalingGroup.InstanceRequirementsProperty
+
   /**
    * The instance type, such as m3.xlarge. You must use an instance type that is supported in your requested Region
    * and Availability Zones.
    *
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts:570]:
```suggestion
   *
   * You must specify one of instanceRequirements or instanceType.
   *
```

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Thanks 👍
I left some notes for minor adjustments.
I think we should require one of the two to be specified, feel free to comment if you think otherwise.

**Comment by wafuwafu13** [comment]
Thank you for reviewing! I fixed.

**Review by sumupitchayan — APPROVED** [review]

**Comment by sumupitchayan** [comment]
@mergifyio update

**Review by sumupitchayan — APPROVED** [review]

**Pull request merged**